src/audio/dai-zephyr.c

```diff
@@ -250,13 +252,69 @@ static enum dma_cb_status dai_dma_cb(struct dai_data *dd, struct comp_dev *dev,
 
 	local_buf = buffer_acquire(dd->local_buffer);
 
-	if (dev->direction == SOF_IPC_STREAM_PLAYBACK)
+	if (dev->direction == SOF_IPC_STREAM_PLAYBACK) {
 		ret = dma_buffer_copy_to(local_buf, dma_buf,
 					 dd->process, bytes);
-	else
-		ret = dma_buffer_copy_from(dma_buf, local_buf,
-					   dd->process, bytes);
+		buffer_release(local_buf);
+	} else {
+		struct list_item *sink_list;
+
+		audio_stream_invalidate(&dma_buf->stream, bytes);
+		/*
+		 * The PCM converter functions used during DMA buffer copy can never fail,
+		 * so no need to check the return value of dma_buffer_copy_from_no_consume().
+		 */
+		ret = dma_buffer_copy_from_no_consume(dma_buf, local_buf, dd->process, bytes);
+		buffer_release(local_buf);
+#if CONFIG_IPC_MAJOR_4
+		/* Skip in case of endpoint DAI devices created by the copier */
+		if (converter) {
+			/*
+			 * copy from DMA buffer to all sink buffers using the right PCM converter
+			 * function
+			 */
+			list_for_item(sink_list, &dev->bsink_list) {
+				struct comp_buffer __sparse_cache *sink_c;
+				struct comp_dev *sink_dev;
+				struct comp_buffer *sink;
+				int j;
+
+				sink = container_of(sink_list, struct comp_buffer, source_list);
+
+				/* this has been handled above already */
+				if (sink == dd->local_buffer)
+					continue;
 
+				sink_c = buffer_acquire(sink);
+				sink_dev = sink_c->sink;
+
+				j = IPC4_SINK_QUEUE_ID(sink_c->id);
+
+				if (j >= IPC4_COPIER_MODULE_OUTPUT_PINS_COUNT) {
+					comp_err(dev, "Sink queue ID: %d >= max output pin count: %d\n",
+						 j, IPC4_COPIER_MODULE_OUTPUT_PINS_COUNT);
+					ret = -EINVAL;
+					goto err;
+				}
+
+				if (!converter[j]) {
+					comp_err(dev, "No PCM converter for sink queue %d\n", j);
+					ret = -EINVAL;
+					goto err;
+				}
+
+				if (sink_dev && sink_dev->state == COMP_STATE_ACTIVE)
```

Comment:
ehm... `if (ret >= 0 && sink_dev && sink_dev->state == COMP_STATE_ACTIVE)` and remove those `goto`s...